builtin/aws/ecs/platform.go

```diff
@@ -1127,305 +1439,756 @@ func (p *Platform) Launch(
+	s.Update("Attempting to find an existing role named %q", roleName)
 
-	dep := &Deployment{
-		Cluster:    clusterName,
-		TaskArn:    taskArn,
-		ServiceArn: *servOut.Service.ServiceArn,
+	queryInput := &iam.GetRoleInput{
+		RoleName: aws.String(roleName),
 	}
 
-	// the TargetGroupArn set here is used by Releaser to set the active
-	// TargetGroup's weight to 100
-	if !p.config.DisableALB {
-		dep.TargetGroupArn = *tgArn
-		dep.LoadBalancerArn = *lbArn
+	svc := iam.New(sess)
+	getOut, err := svc.GetRoleWithContext(ctx, queryInput)
+	if err != nil {
+		return fmt.Errorf("requested task IAM role not found: %s", roleName)
 	}
 
-	return dep, nil
+	s.Update("Found existing task IAM role: %s", roleName)
+
+	state.Name = roleName
+	state.Arn = *getOut.Role.Arn
+	s.Done()
+	return nil
 }
 
-func destroyALB(
+func (p *Platform) resourceExecutionRoleCreate(
 	ctx context.Context,
+	sg terminal.StepGroup,
 	log hclog.Logger,
 	sess *session.Session,
-	deployment *Deployment,
+	src *component.Source,
+	state *Resource_ExecutionRole,
 ) error {
-	log.Debug("removing deployment target group from load balancer")
+	s := sg.Add("Initiating execution role creation...")
+	defer s.Abort()
 
-	elbsrv := elbv2.New(sess)
-
-	log.Debug("load balancer arn", "arn", deployment.LoadBalancerArn)
+	roleName := p.config.ExecutionRoleName
 
-	listeners, err := elbsrv.DescribeListeners(&elbv2.DescribeListenersInput{
-		LoadBalancerArn: &deployment.LoadBalancerArn,
-	})
-	if err != nil {
-		return err
+	if roleName == "" {
+		roleName = "ecr-" + src.App
+		state.Managed = true
+	} else {
+		// If the role name is defined, we're not managing this role, and shouldn't destroy it later.
+		state.Managed = false
 	}
+	// role names have to be 64 characters or less, and the client side doesn't validate this.
+	if len(roleName) > 64 {
+		roleName = roleName[:64]
+		log.Debug("using a shortened value for role name due to AWS's length limits", "roleName", roleName)
+	}
+	state.Name = roleName
 
-	var listener *elbv2.Listener
-
-	if len(listeners.Listeners) > 0 {
-		listener = listeners.Listeners[0]
-
-		log.Debug("listener arn", "arn", *listener.ListenerArn)
-
-		def := listener.DefaultActions
+	s.Update("Attempting to find an existing role named %q", roleName)
 
-		var tgs []*elbv2.TargetGroupTuple
+	queryInput := &iam.GetRoleInput{
+		RoleName: aws.String(roleName),
+	}
 
-		// If there is only 1 target group, delete the listener
-		if len(def) == 1 && len(def[0].ForwardConfig.TargetGroups) == 1 {
-			log.Debug("only 1 target group, deleting listener")
-			_, err = elbsrv.DeleteListener(&elbv2.DeleteListenerInput{
-				ListenerArn: listener.ListenerArn,
-			})
+	svc := iam.New(sess)
 
-			if err != nil {
-				return err
-			}
-		} else if len(def) > 0 && def[0].ForwardConfig != nil && len(def[0].ForwardConfig.TargetGroups) > 1 {
-			// Multiple target groups means we can keep the listener
-			var active bool
+	getOut, err := svc.GetRoleWithContext(ctx, queryInput)
```

Comment:
With the improved resource visibility, we can see that _this_ turns out to be the slowest part of the ecs deployment most of the time. Go figure.